File: Python3/프로그래머스/2/42577. 전화번호 목록/전화번호 목록.py
def solution(phone_book):
    # in 연산자는 set/dict은 O(1), list/tuple은 O(N)
    prefixs = set(phone_book)
    for number in phone_book:
        # 선택한 번호에 대해서 하나씩 증가하며 접두사 생성
        for i in range(1, len(number)):
            prefix = number[:i]
            # 접두사가 있으면 종료
            if prefix in prefixs:
                return False
    return True

File: C++/프로그래머스/2/42577. 전화번호 목록/전화번호 목록.cpp
#include <string>    
#include <vector>      
#include <algorithm>  

using namespace std;

// 전화번호 목록에서 접두어 관계가 있는지 확인하는 함수
bool solution(vector<string> phone_book) {
    // 1. 전화번호 목록을 사전순으로 정렬합니다.
    sort(phone_book.begin(), phone_book.end());

    // 2. 정렬된 전화번호 목록에서 인접한 전화번호끼리 접두어 관계인지 확인합니다.
    for(int i = 0; i < phone_book.size() - 1; i++) {
        // 0이 반환된 경우, 접두어로 판단
        if(phone_book[i + 1].find(phone_book[i]) == 0) {
            return false; 
        }
    }

    // 3. 모든 전화번호를 비교했을 때 접두어 관계가 없으면 true를 반환합니다.
    return true;
}
